File: python/01_basic_chat.py
#!/usr/bin/env python3
"""
Example 1: Basic Chat
This sends a simple question to the AI model using Python's built-in libraries.
No external dependencies needed (no pip install)!
"""

import http.client
import json
import os

# Get API key from environment variable
api_key = os.environ.get('GROQ_API_KEY')

if not api_key:
    print("Error: GROQ_API_KEY environment variable not set")
    print("Run: export GROQ_API_KEY='your_key_here'")
    exit(1)

# Create the request payload (the data we're sending)
payload = {
    "model": "meta-llama/llama-4-scout-17b-16e-instruct",
    "messages": [
        {
            "role": "user",
            "content": "What is the capital of Switzerland?"
        }
    ],
    "temperature": 0.7,
    "max_tokens": 100
}

# Convert Python dict to JSON string
payload_json = json.dumps(payload)

# Create HTTPS connection
conn = http.client.HTTPSConnection("api.groq.com")

# Set up headers
headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {api_key}'
}

# Make the request
conn.request("POST", "/openai/v1/chat/completions", payload_json, headers)

# Get the response
response = conn.getresponse()

# Read and parse the response
response_data = response.read().decode('utf-8')
response_json = json.loads(response_data)

# Close the connection
conn.close()

# Display the results
print("Full Response:")
print(json.dumps(response_json, indent=2))

print("\nAI Answer:")
print(response_json['choices'][0]['message']['content'])

print("\nToken Usage:")
print(f"  Prompt: {response_json['usage']['prompt_tokens']}")
print(f"  Response: {response_json['usage']['completion_tokens']}")
print(f"  Total: {response_json['usage']['total_tokens']}")

# Python concepts explained:
#
# import http.client
#   Built-in module for HTTP/HTTPS requests (no pip install!)
#
# import json
#   Built-in module for JSON encoding/decoding
#
# import os
#   Built-in module for OS operations (like environment variables)
#
# os.environ.get('VAR_NAME')
#   Gets environment variable (like $GROQ_API_KEY in bash)
#   Returns None if not set
#
# json.dumps(dict)
#   Converts Python dictionary to JSON string
#   indent=2 means pretty-print with 2-space indentation
#
# json.loads(string)
#   Converts JSON string to Python dictionary
#
# f'Bearer {api_key}'
#   F-string (formatted string literal)
#   Variables in {curly braces} get replaced with their values
#
# http.client.HTTPSConnection()
#   Creates HTTPS connection (secure HTTP)
#   Use HTTPConnection() for plain HTTP (not recommended)
#
# conn.request(method, path, body, headers)
#   Sends HTTP request
#   method = "GET", "POST", "PUT", etc.
#   path = "/api/endpoint"
#   body = request data (JSON string)
#   headers = dict of HTTP headers
#
# response.read()
#   Returns bytes (binary data)
#   Need to .decode('utf-8') to get string
#
# Dictionary access:
#   response_json['key']           # Crashes if key missing
#   response_json.get('key')       # Returns None if key missing
#   response_json.get('key', 'default')  # Returns 'default' if missing
#
# String methods:
#   .upper()    # "hello" → "HELLO"
#   .lower()    # "HELLO" → "hello"
#   .strip()    # "  hi  " → "hi"
#   .split()    # "a b c" → ['a', 'b', 'c']

File: nodejs/01_basic_chat.js
// Example 1: Basic Chat
// This sends a simple question to the AI model using Node.js built-in https module.
// No external dependencies needed!

const https = require('https');

// The request data (payload) that we'll send to the API
const data = JSON.stringify({
  model: "meta-llama/llama-4-scout-17b-16e-instruct",
  messages: [
    {
      role: "user",
      content: "What is the capital of Switzerland?"
    }
  ],
  temperature: 0.7,
  max_tokens: 100
});

// Configure the HTTP request
const options = {
  hostname: 'api.groq.com',
  port: 443,                                    // HTTPS port
  path: '/openai/v1/chat/completions',
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${process.env.GROQ_API_KEY}`,
    'Content-Length': Buffer.byteLength(data)  // Required for POST
  }
};

// Make the request
const req = https.request(options, (res) => {
  let responseData = '';

  // Receive data in chunks
  res.on('data', (chunk) => {
    responseData += chunk;
  });

  // When all data received
  res.on('end', () => {
    try {
      const parsed = JSON.parse(responseData);

      // Pretty print the full response
      console.log('Full Response:');
      console.log(JSON.stringify(parsed, null, 2));

      // Extract just the answer
      console.log('\nAI Answer:');
      console.log(parsed.choices[0].message.content);

      // Show token usage
      console.log('\nToken Usage:');
      console.log(`Prompt: ${parsed.usage.prompt_tokens}`);
      console.log(`Response: ${parsed.usage.completion_tokens}`);
      console.log(`Total: ${parsed.usage.total_tokens}`);

    } catch (error) {
      console.error('Error parsing response:', error.message);
      console.log('Raw response:', responseData);
    }
  });
});

// Handle request errors
req.on('error', (error) => {
  console.error('Request failed:', error.message);

  if (error.message.includes('ENOTFOUND')) {
    console.log('Check your internet connection');
  }
});

// Send the request with our data
req.write(data);
req.end();

// Explanation of Node.js concepts:
//
// require('https')
//   Imports the built-in HTTPS module (no npm install needed)
//
// JSON.stringify()
//   Converts JavaScript object to JSON string
//   Second parameter (null, 2) = pretty print with 2 spaces
//
// process.env.GROQ_API_KEY
//   Reads environment variable (same as $GROQ_API_KEY in bash)
//
// Buffer.byteLength(data)
//   Calculates exact size of data in bytes
//   Required for the Content-Length header
//
// https.request()
//   Creates an HTTPS request object
//   Returns immediately (doesn't wait for response)
//
// res.on('data', ...)
//   Event fired when data arrives
//   Data comes in "chunks" (pieces), not all at once
//
// res.on('end', ...)
//   Event fired when ALL data has arrived
//   Now we can process the complete response
//
// req.write(data)
//   Sends our JSON data to the server
//
// req.end()
//   Tells server we're done sending data
//   Without this, request will hang forever!
//
// Event-driven programming:
//   Node.js doesn't wait for responses
//   Instead, you provide callback functions
//   These run when events happen (data arrives, request ends, etc.)
